### ISSUE TITLE
callback bug

### ISSUE THREAD

**Issue by kaiko** [Marak/node_mailer#15]
Hello.

You have bug: you cache send command callback.
If I'll do:

```
function s(callback) {
  email.send({to:..., .... }, callback);
}

s(function () { util.puts("a") });
s(function () { util.puts("b") });
```

... then I'll get "a" twice because you cache callback to connection.

**Comment by Marak** [comment]
That shouldn't happen, I'll look into this asap.

**Comment by kaiko** [comment]
My fast fix was in Email function:

```
var c = new Connection(options, function () {
   c.disconnect();
   if (callback) callback();
});
c.queue(this);
```

This creates new connection every time which is not good but better than same callback called all the time.

**Comment by Marak** [comment]
Well, shouldn't there only be one error thrown anyway per connection? I think there is a gap in the api.

**Comment by kaiko** [comment]
But I want to know also if it was sent successfully.
If you do:
    m1 = {...};
    m2 = {...};
    m3 = {...}
    function emailerror(err) { util.puts(err) });
    email.send(m1, emailerror);
    email.send(m2, emailerror); // ignored callback argument, can't use closures
    email.send(m3, emailerror); // ignored callback argument, can't use closures

How do you know which mail is already sent and which is not?
If you want to send 1000 e-mails and suddenly something happen then you have no idea which emails you have to repeat and which not.

If you planned already this way then I think this is logic bug and a lot of people can't use this library in critical situations.

**Comment by Marak** [comment]
I'm starting to remember now. Mail needs to be an EventEmitter then emits a success / fail event after each email has been processed. 

I started to implement this in a new branch but never got around to it. I suppose making the event emitter is the right API choice here.

**Comment by Marak** [comment]
maybe something like, 

```
    var email = require("node_mailer");

    email.on('send', function(){});
    email.on('error', function(){});

    email.send({});
    email.send({});
    email.send({});
```

Does that make sense?

**Comment by kaiko** [comment]
If you pass send function argument object 1:1 to event then it helps.
It gives a possibility that I pass my own ID for example and then read it in event functions. Like
    var email = require("node_mailer");
    email.on('send', function (m) {
      util.puts("Email " + m.id + " is sent to " + m.to)
    })
    email.send({id:1,
       from: "Kaiko Kaur kaiko.kaur@gmail.com",
       to: "marak.squires@gmail.com", ...
    )};
    email.send({id: 2, ... })

**Comment by jaekwon** [comment]
ping

**Comment by dachev** [comment]
ping

**Comment by Marak** [comment]
I'm sorry, I don't have time to fix this issue / maintain this part of the library. 

I would either suggest looking at some of the other node.js mailer libraries ( there are plenty ) , or attempt to the patch the issue. I'd be glad to add anyone as repo admin.

Sorry I can't be of more help.

**Comment by dachev** [comment]
Fair enough. I'll look into it if I get time on the weekend.

Blago

**Comment by jaekwon** [comment]
fyi the nodemailer library is working alright for me.

**Comment by Marak** [comment]
This should be resolved now with the version 0.5.0 and the new smtp library. Can anyone confirm?
